### PR TITLE
[AMDGPU][GFX12] Add Atomic cond_sub_u32

### PR DESCRIPTION
**Pull request by mariusz-sikora-at-amd**

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 1]
```diff
@@ -2491,6 +2499,11 @@ def int_amdgcn_flat_atomic_fmax_num   : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 def int_amdgcn_global_atomic_fmin_num : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 def int_amdgcn_global_atomic_fmax_num : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 
+def int_amdgcn_flat_atomic_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, flat_ptr_ty>;
+def int_amdgcn_global_atomic_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, global_ptr_ty>;
+
+def int_amdgcn_ds_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, local_ptr_ty>;
```
Comment (arsenm) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2505]: The other memory operations we just have one intrinsic mangled by the pointer type. We should also consider just adding it to atomicrmw

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td [hunk 2]
```diff
@@ -2502,10 +2500,9 @@ def int_amdgcn_flat_atomic_fmax_num   : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 def int_amdgcn_global_atomic_fmin_num : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 def int_amdgcn_global_atomic_fmax_num : AMDGPUAtomicRtn<llvm_anyfloat_ty>;
 
-def int_amdgcn_flat_atomic_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, flat_ptr_ty>;
-def int_amdgcn_global_atomic_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, global_ptr_ty>;
-
-def int_amdgcn_ds_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty, local_ptr_ty>;
+def int_amdgcn_flat_atomic_cond_sub_u32   : AMDGPUAtomicRtn<llvm_i32_ty>;
+def int_amdgcn_global_atomic_cond_sub_u32 : AMDGPUAtomicRtn<llvm_i32_ty>;
+def int_amdgcn_ds_cond_sub_u32            : AMDGPUAtomicRtn<llvm_i32_ty>;
```
Comment (mariusz-sikora-at-amd) [inline llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2505]: @arsenm  is this what you were expecting ?

**Comment by mariusz-sikora-at-amd** [comment]
Adding support in atomicrmw. This will require to add new operation to aromicrmw "cond_sub" or you had something else in mind @arsenm ?

**Comment by mariusz-sikora-at-amd** [comment]
ping

**Comment by jayfoad** [comment]
> Adding support in atomicrmw. This will require to add new operation to aromicrmw "cond_sub"

Yes, and we have (Matt has) done this in the past, but it will require a wider consensus. I think it's fine to add AMDGPU intrinsics for this in the mean time.

**Review by arsenm — COMMENTED** [review]
Missing UniformityAnalysis test for these

**Comment by mariusz-sikora-at-amd** [comment]
> Missing UniformityAnalysis test for these

Done

File: llvm/include/llvm/IR/IntrinsicsAMDGPU.td (hunk 2 above)
Comment (arsenm) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td:2505]: No, I meant one intrinsic, not 3. When we had atomicrmw inc/dec intrinsics, we had one intrinsic per operation overloaded per address space

Comment (mariusz-sikora-at-amd) [reply llvm/include/llvm/IR/IntrinsicsAMDGPU.td]: done

File: llvm/lib/Target/AMDGPU/AMDGPUGISel.td [hunk 3]
```diff
@@ -264,6 +264,7 @@ def : GINodeEquiv<G_AMDGPU_BUFFER_ATOMIC_FADD, SIbuffer_atomic_fadd>;
 def : GINodeEquiv<G_AMDGPU_BUFFER_ATOMIC_FMIN, SIbuffer_atomic_fmin>;
 def : GINodeEquiv<G_AMDGPU_BUFFER_ATOMIC_FMAX, SIbuffer_atomic_fmax>;
 def : GINodeEquiv<G_AMDGPU_BUFFER_ATOMIC_CMPSWAP, SIbuffer_atomic_cmpswap>;
+def : GINodeEquiv<G_AMDGPU_BUFFER_ATOMIC_COND_SUB_U32, SIbuffer_atomic_cond_sub_u32>;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUGISel.td:267]: Indentation is off

**Review by arsenm — APPROVED** [review]
Maybe add references in AMDGPUUsage intrinsics section

File: llvm/docs/AMDGPUUsage.rst [hunk 4]
```diff
@@ -1182,6 +1182,11 @@ The AMDGPU backend implements the following LLVM IR intrinsics.
 
                                                    The iglp_opt strategy implementations are subject to change.
 
+  llvm.atomic.cond.sub.u32                         Provides direct access to flat_atomic_cond_sub_u32, global_atomic_cond_sub_u32
```
Comment (arsenm) [inline llvm/docs/AMDGPUUsage.rst:1185]: missed the amdgcn part

**Review by arsenm — APPROVED** [review]

**Pull request merged**